### PR TITLE
schedule: fix a thread-safe bug and improve code

### PR DESCRIPTION
**Pull request by Luffbee**
### What problem does this PR solve? 
1. There is a thread-safe bug which may remove operators unexpectedly, see issue #1716.
2. There is a suspicious code in `pollNeedDispathOperator`: when the region of an operator is disappeared, it is ignored. The code to handle this problem is in `RaftCluster.checkOperator`, which is far away from the problem.

### What is changed and how it works?
1. This PR use the second possible solution mentioned in the issue: check operator is the same before `delete`.
2. Moved the logic in `RaftCluster.checkOperator` to `pollNeedDispathOperator`, and removed some useless functions.

### Check List 

Tests 

 - Unit test

### CODE REVIEW

File: server/schedule/operator_controller_test.go [hunk 1]
```diff
@@ -98,27 +101,75 @@ func (t *testOperatorControllerSuite) TestOperatorStatus(c *C) {
...
+		operator.AddPeer{ToStore: 1, PeerID: 4},
+	}
+	// finished op with normal priority
+	op1 := operator.NewOperator("test", "test", 1, &metapb.RegionEpoch{}, operator.OpRegion, operator.TransferLeader{ToStore: 2})
+	// unfinished op with high priority
+	op2 := operator.NewOperator("test", "test", 1, &metapb.RegionEpoch{}, operator.OpRegion|operator.OpAdmin, steps...)
+
+	oc.SetOperator(op1)
+
+	c.Assert(failpoint.Enable("github.com/pingcap/pd/server/schedule/concurrentRemoveOperator", "return(true)"), IsNil)
```
Comment (nolouch) [inline server/schedule/operator_controller_test.go]: Can we stably reproduce the problem in the issue? maybe we can use `pause` on failpoint. likes https://github.com/pingcap/pd/blob/master/tests/server/watch/leader_watch_test.go#L62

File: server/schedule/operator_controller.go [hunk 2]
```diff
@@ -306,10 +318,10 @@ func (oc *OperatorController) addOperatorLocked(op *operator.Operator) bool {
 	// If there is an old operator, replace it. The priority should be checked
 	// already.
 	if old, ok := oc.operators[regionID]; ok {
+		_ = oc.removeOperatorLocked(old)
```
Comment (rleungx) [inline server/schedule/operator_controller.go]: Maybe we should handle this value?

File: server/schedule/operator_controller_test.go (hunk 1 above)
Comment (Luffbee) [reply server/schedule/operator_controller_test.go]: I have already reproduce the problem in issue. Remove the check in `removeOperatorLocked` and enable failpoint, this test will fail.

File: server/schedule/operator_controller.go (hunk 2 above)
Comment (Luffbee) [reply server/schedule/operator_controller.go]: I don't think so, this function already get the lock, so we can be sure `oc.operators` won't change between line 320 and 321. The return value must be true.

File: server/schedule/operator_controller_test.go (hunk 1 above)
Comment (nolouch) [reply server/schedule/operator_controller_test.go]: got it!

**Comment by nolouch** [comment]
PTAL @rleungx @shafreeck

File: server/handler.go [hunk 3]
```diff
@@ -304,7 +304,7 @@ func (h *Handler) RemoveOperator(regionID uint64) error {
 		return ErrOperatorNotFound
 	}
 
-	c.opController.RemoveOperator(op)
+	_ = c.opController.RemoveOperator(op)
```
Comment (rleungx) [inline server/handler.go]: How about this one, should we handle it?

Comment (Luffbee) [reply server/handler.go]: I think the aim of `Handle.RemoveOperator` is to remove the `op`. If this one returns false, the `op` must be removed by another thread. So, the aim will be achieved no matter what is returned.

Comment (rleungx) [reply server/handler.go]: It seems that we can remove all `_` when we don't need to handle it.

Comment (Luffbee) [reply server/handler.go]: I'd like to use the `_` to indicate that the return value is ignored intentionally.

File: server/schedule/operator_controller.go [hunk 4]
```diff
@@ -367,11 +371,15 @@ func (oc *OperatorController) GetOperatorStatus(id uint64) *OperatorWithStatus {
 	return oc.opRecords.Get(id)
 }
 
-func (oc *OperatorController) removeOperatorLocked(op *operator.Operator) {
+func (oc *OperatorController) removeOperatorLocked(op *operator.Operator) bool {
 	regionID := op.RegionID()
-	delete(oc.operators, regionID)
-	oc.updateCounts(oc.operators)
-	operatorCounter.WithLabelValues(op.Desc(), "remove").Inc()
+	if cur := oc.operators[regionID]; cur == op {
```
Comment (shafreeck) [inline server/schedule/operator_controller.go]: I understand the bug here caused by different operators having the same region ID.  The check here does not guarantees deleting the correct operator, I think.

Comment (Luffbee) [reply server/schedule/operator_controller.go]: Here is comparing two pointers. Why cannot it guarantee deleting the correct operator? Could you give an example?

Comment (shafreeck) [reply server/schedule/operator_controller.go]: My mistake, I didn't notice the comparison. I thought it just judges the existence.

**Comment by rleungx** [comment]
@Luffbee Please solve the conflicts.

**Pull request merged**